### PR TITLE
Add vertex texture buffers for VS input and support attribute instancing using VertexID

### PR DESCRIPTION
**Pull request by kd-11**
Fix vertex decompilation and attribute location

Add vertex texture upload for missing case

Add frequency scaling to allow for vertex instancing (Resogun)

Clean up the code changes for submission

### CODE REVIEW

File: rpcs3/Emu/RSX/Common/BufferUtils.cpp [hunk 1]
```diff
@@ -62,7 +62,7 @@ void write_vertex_array_data_to_buffer(void *buffer, u32 first, u32 count, size_
 				*c_dst++ = *c_src++;
 			}
 			if (vertex_array_desc.size * sizeof(u16) < element_size)
-				*c_dst++ = 0x3800;
+				*c_dst++ = 0x3c00;		//was 0x3800=0.5; why 0.5, should'nt this have always been 1.0??
```
Comment (tambry) [inline rpcs3/Emu/RSX/Common/BufferUtils.cpp]: Why does the comment have 2 tabs before it?

Comment (danilaml) [inline rpcs3/Emu/RSX/Common/BufferUtils.cpp]: @vlj added this line, so he could probably answer.

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp [hunk 2]
```diff
@@ -30,6 +30,25 @@ namespace
...
+		 * This rules out most 3-component formats, except for the 32-wide RGB32F, RGB32I, RGB32UI
+		 */
+		const u32 vec1_types[] = { GL_R16, GL_R32F, GL_R16F, GL_R8, GL_R32I, GL_R16F, GL_R8 };
+		const u32 vec2_types[] = { GL_RG16, GL_RG32F, GL_RG16F, GL_RG8, GL_RG32I, GL_RG16F, GL_RG8 };
+		const u32 vec3_types[] = { GL_RGBA16, GL_RGB32F, GL_RGBA16F, GL_RGBA8, GL_RGB32I, GL_RGBA16F, GL_RGBA8 };	//VEC3 COMPONENTS NOT SUPPORTED!
+		const u32 vec4_types[] = { GL_RGBA16, GL_RGBA32F, GL_RGBA16F, GL_RGBA8, GL_RGBA32I, GL_RGBA16F, GL_RGBA8 };
+
+		const u32* vec_selectors[] = { 0, vec1_types, vec2_types, vec3_types, vec4_types };
+
+		if ((int)type > 6)
```
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: Seems hacky. Why > 6? If it's some enum value, why not compare with it?

File: rpcs3/Emu/RSX/GL/GLVertexProgram.cpp [hunk 3]
```diff
@@ -36,10 +36,33 @@ void GLVertexDecompilerThread::insertHeader(std::stringstream &OS)
...
+	 * Its is important that the locations are in the order that vertex attributes are expected.
+	 * If order is not adhered to, channels may be swapped leading to corruption
+	*/
+
+	std::sort(input_data.begin(), input_data.end());
+
+	int location = 1;
+	for (const std::tuple<size_t, std::string> item : input_data)
+	{
+		for (const ParamType PT : inputs)
```
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/GLVertexProgram.cpp]: Why not `const ParamType &PT` ?

File: rpcs3/Emu/RSX/Common/BufferUtils.cpp (hunk 1 above)
Comment (kd-11) [inline rpcs3/Emu/RSX/Common/BufferUtils.cpp]: The comment is temporary; to be removed once vlj clears things up

**Comment by kd-11** [comment]
@danilaml Fixed the missing ref and enum comparison

**Comment by tambry** [comment]
@kd-11 While close/reopen the PR?

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp [hunk 4]
```diff
@@ -30,6 +30,25 @@ namespace
...
+		 * This rules out most 3-component formats, except for the 32-wide RGB32F, RGB32I, RGB32UI
+		 */
+		const u32 vec1_types[] = { GL_R16, GL_R32F, GL_R16F, GL_R8, GL_R32I, GL_R16F, GL_R8 };
+		const u32 vec2_types[] = { GL_RG16, GL_RG32F, GL_RG16F, GL_RG8, GL_RG32I, GL_RG16F, GL_RG8 };
+		const u32 vec3_types[] = { GL_RGBA16, GL_RGB32F, GL_RGBA16F, GL_RGBA8, GL_RGB32I, GL_RGBA16F, GL_RGBA8 };	//VEC3 COMPONENTS NOT SUPPORTED!
+		const u32 vec4_types[] = { GL_RGBA16, GL_RGBA32F, GL_RGBA16F, GL_RGBA8, GL_RGBA32I, GL_RGBA16F, GL_RGBA8 };
+
+		const u32* vec_selectors[] = { 0, vec1_types, vec2_types, vec3_types, vec4_types };
+
+		if (type > rsx::vertex_base_type::ub256)
```
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: Is this possible though? ub256 is the latest in the enum.

**Comment by kd-11** [comment]
@tambry sorry for that, mis-clicked the comment button

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp (hunk 4 above)
Comment (kd-11) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: If anything else was to be added, it'd probably segfault without the check. I could remove it if seems unnecessary

Comment (tambry) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: @kd-11 You should probably throw an exception if that happens, instead of silently failing.

**Comment by kd-11** [comment]
travis ci fails error: ‘PFNGLTEXTUREBUFFERPROC’ does not name a type. Any hints? I declared some required gl functions in glProcTable. Is there any other file that needs changing?

**Comment by danilaml** [comment]
Maybe @DHrpcs3 or @vlj can help with that.

**Comment by vlj** [comment]
It's strange. Do we use system gl.h / glext.h ?
It's possible that the function is already defined/available without glProcTable on linux.

**Comment by danilaml** [comment]
@vlj we have glext.h in our repo (updated) but I'm not sure if we use it on linux.

**Comment by vlj** [comment]
btw thanks for taking care of the GLbackend.

**Comment by kd-11** [comment]
@vlj its no problem really. i've been using the emu for years without really contributing, so I thought i'd start.

**Comment by kd-11** [comment]
Is it possible to take a look at the glew header used in the travis ci unix environments? it builds fine on mac.

**Comment by danilaml** [comment]
@kd-11 it installs http://packages.ubuntu.com/precise/libglew1.6-dev

**Comment by danilaml** [comment]
Try using libglew from `debian-sid` ppa (I think if you just add it to apt sources list it'd fetch latest glew from there automatically)

**Comment by kd-11** [comment]
You're right, glew 1.6 only has the EXT versions of those functions. Should I edit the travis.yml directly or is this task left to maintainers?

**Comment by tambry** [comment]
@kd-11 It's OK for you to edit the Travis build file.

**Comment by kd-11** [comment]
debian-sid causes a conflict, but actually has an updated glew version..
dpkg: error processing /var/cache/apt/archives/locales_2.21-7_all.deb (--unpack):
 trying to overwrite '/usr/sbin/validlocale', which is also in package libc-bin 2.15-0ubuntu10.12
dpkg-deb (subprocess): subprocess data was killed by signal (Broken pipe)
dpkg-deb: error: subprocess <decompress> returned error exit status 2

I haven't used apt in a while, so anyone with advice on how to get around this?

File: rpcs3/Emu/RSX/GL/GLVertexProgram.cpp [hunk 5]
```diff
@@ -101,6 +124,37 @@ void GLVertexDecompilerThread::insertOutputs(std::stringstream & OS, const std::
 	}
 }
 
+void add_input(std::stringstream & OS, const ParamItem &PI, const std::vector<rsx_vertex_input> &inputs)
```
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/GLVertexProgram.cpp]: Small style inconsistency here: `std::stringstream & OS`

Comment (tambry) [inline rpcs3/Emu/RSX/GL/GLVertexProgram.cpp]: Possibly a copy-paste error, as the new argument has the & properly formatted.

File: rpcs3/Emu/RSX/GL/gl_helpers.h [hunk 6]
```diff
@@ -1131,6 +1132,33 @@ namespace gl
 			__glcheck glTexSubImage2D((GLenum)get_target(), level(), 0, 0, width(), height(), (GLenum)format, (GLenum)type, src);
 		}
 
+		void copy_from(buffer buf, u32 gl_format_type, u32 offset, u32 length)
+		{
+			if (get_target() != target::textureBuffer)
+				throw;
```
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/gl_helpers.h]: Why?

Comment (tambry) [inline rpcs3/Emu/RSX/GL/gl_helpers.h]: May well throw, but at least throw EXCEPTION and add to it what went wrong.

**Comment by danilaml** [comment]
@kd-11 what version of oGL do you use? It seems that `PFNGLTEXTUREBUFFERPROC` is found only in oGL 4.5 (at least not as extension).

**Comment by kd-11** [comment]
@danilaml you're right, the others are in core 3.1 but TextureBuffer is in 4.5. Guess i'll make it work with TexBuffer then

**Comment by kd-11** [comment]
Still failing due to TexBufferRange which is core since 4.3. Buffer textures in the form used here can only be properly accessed using glsl version 430 in the VS anyway.

**Comment by kd-11** [comment]
Reworked to base on OpenGL 3.1 with EXT extensions. However, glew header is still too old and does not have definitions for some newer extensions.

**Comment by danilaml** [comment]
@kd-11 you shouldn't have. 4.3 Is fine, I think.  https://github.com/danilaml/rpcs3/commits/patch-2

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp [hunk 7]
```diff
@@ -30,6 +30,25 @@ namespace
...
+			throw std::runtime_error(fmt::format("OpenGL error: unknown vertex base type 0x%X. file '%s' function '%s' line %ld", (u32)type, __FILE__, __FUNCTION__, __LINE__));
```
Comment (tambry) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: std::runtime_error? Why not use the standard EXCEPTION, that we use in our codebase.
For example:
throw EXCEPTION("Unknown vertex base type: 0x%x", (u32)type);

It also logs the namespace, function and the line number already.

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp (hunk 4 above)
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: Could something be added though? @vlj any thought?
Maybe dummy last element could be added so we can static_assert that it is 7.

**Comment by kd-11** [comment]
@danilaml The header in travis-ci is from several years back; 2011 if i'm not mistaken. There are alot of missing definitions there.

**Comment by tambry** [comment]
@kd-11 You could keep it for OpenGL 4.3 probably, as @danilaml said. You can probably also use the workaround, that he posted.

**Comment by danilaml** [comment]
@kd-11 yes, I upgraded libglew to 1.10 on top of your previous changes and it compiled fine: https://github.com/danilaml/rpcs3/commits/patch-2

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp (hunk 4 above)
Comment (vlj) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: If type is above ub256 then it's likely the convert function has throw an exception before to_gl_internal_type is reached. It's not possible (except with raw cast) but maybe MSVC will emit a warning if there's no check.

**Comment by tambry** [comment]
@kd-11 Both your PRs require a rebase.

**Comment by kd-11** [comment]
rebased

**Comment by danilaml** [comment]
@kd-11 there are still concerns regarding https://github.com/RPCS3/rpcs3/pull/1461#discussion-diff-51289663R47
(why std exception instead of `EXCEPTION`)?

**Comment by kd-11** [comment]
@danilaml I converted them a short while ago.

File: rpcs3/Emu/RSX/GL/gl_helpers.h [hunk 8]
```diff
@@ -1131,6 +1133,32 @@ namespace gl
...
+			if (!offset)
+			{
+				copy_from(buf, gl_format_type);
+				return;
+			}
+
+			if (glTextureBufferRangeEXT == nullptr)
+			{
+				throw EXCEPTION("OpenGL error: partial buffer access for textures is unsupported on your system");
+				return;
```
Comment (tambry) [inline rpcs3/Emu/RSX/GL/gl_helpers.h]: Exceptions don't return.

Comment (kd-11) [inline rpcs3/Emu/RSX/GL/gl_helpers.h]: fixed

**Comment by kd-11** [comment]
updated to fix a crash in outrun in the case of inlined arrays

**Comment by kd-11** [comment]
trigger travis rebuild

**Comment by kd-11** [comment]
Rebased.

This PR exposes this issue https://github.com/RPCS3/rpcs3/issues/1471 that specifically affects AMD drivers since they tend to be much stricter than NV

**Comment by kd-11** [comment]
Trigger rebuild

**Comment by kd-11** [comment]
Appveyor fails without actually giving any error message.

File: rpcs3/Emu/RSX/Common/BufferUtils.cpp [hunk 9]
```diff
@@ -62,7 +62,7 @@ void write_vertex_array_data_to_buffer(void *buffer, u32 first, u32 count, size_
 				*c_dst++ = *c_src++;
 			}
 			if (vertex_array_desc.size * sizeof(u16) < element_size)
-				*c_dst++ = 0x3800;
+				*c_dst++ = 0x3c00;
```
Comment (danilaml) [inline rpcs3/Emu/RSX/Common/BufferUtils.cpp]: I get that It was confirmed to be correct?

**Comment by danilaml** [comment]
@kd-11 I triggered re-build, it went fine. Could be ransom OoM.

File: rpcs3/Emu/RSX/Common/BufferUtils.cpp (hunk 9 above)
Comment (kd-11) [inline rpcs3/Emu/RSX/Common/BufferUtils.cpp]: @danilaml Thanks for the rebuild. I cleared up the issue with the constant with vlj. The change is indeed correct.

**Comment by danilaml** [comment]
@vlj it breaks oGL renderer for amd and probably for intel gfx as mentioned in here: https://github.com/RPCS3/rpcs3/issues/1471

**Pull request merged**